### ISSUE TITLE
[12.0.0+] Incorrect date format conversion on input

### ISSUE THREAD

**Issue by adrianszymanski89** [handsontable/handsontable#9681]
### Description

When we have `dateFromat` set to `MM/DD/YY` and we input, for example - 7/1/22 in the cell, it is incorrectly converted to `01/07/2022`. It used to be converted correctly in Handsontable v. 11.1.0 (before updating moment.js dependency #9381)

### Steps to reproduce

1. Type 7/1/22 in the date cell;
2. Press enter;

### Demo

V 11.1.0 (works correctly) - https://jsfiddle.net/aszymanski/nqv3btc0/1/
V 12.0.0 (and higher, it's broken) - https://jsfiddle.net/aszymanski/nqv3btc0/2/

### Your environment
* Handsontable version: 12.0.0+
* Browser name and version: Google Chrome 103
* Operating system: macOS Monterey 12.4

**Comment by adrianszymanski89** [comment]
Inform ZD (29687)

**Comment by aninde** [comment]
Another example for this case
date format: 'YYYY-MM-DD'
Typing `10-11-12`, I expected 2010-11-12, but the result is inverted to 2012-11-10.

https://user-images.githubusercontent.com/30186818/179528805-a742af21-6b6c-4f59-bd04-d1a300a07847.mov

**Comment by wszymanski** [comment]
~~What is the goal? If you want only one digit of the date to be displayed (if you are dealing with a single digit number) please use appropriate `M/D/YYYY` format. Please let me know if that's what you mean.~~

edit: Okey, it's probably about `correctFormat`. Please provide information about it in the issue description and additionally show an expected result.

I've checked the code and it seems that used `correctFormat` method hasn't changed.

https://github.com/handsontable/handsontable/blob/f4c2ca08f578b87177816614dfad68864d81f887/src/validators/dateValidator.js#L46-L62

Thus, if it's actually a regression then upgraded `moment.js` may have bug or library [API has changed](https://github.com/moment/moment/blob/develop/CHANGELOG.md). We upgraded from `2.24.0` to `2.29.3` (12.1.0 release) within https://github.com/handsontable/handsontable/pull/9382 and from `2.29.3` to `2.29.4` (12.1.2 release) within #9637.

**Comment by warpech** [comment]
> Okey, it's probably about correctFormat. Please provide information about it in the issue description and additionally show an expected result.

I don't think so. I can see the same problem without `correctFormat`:

- 11.1 https://jsfiddle.net/warpech/wa58goL2/1/
- 12.0 https://jsfiddle.net/warpech/xjrf0y1p/1/

I think the problem is that Pikaday transforms the input before closing the editor. On this line, Pikaday chooses `defaultDateFormat`:

https://github.com/handsontable/handsontable/blob/6a1706061163b8d1f7752e54ef6f10efbc764b8b/handsontable/src/editors/dateEditor/dateEditor.js#L242

**Comment by wszymanski** [comment]
> > Okey, it's probably about correctFormat. Please provide information about it in the issue description and additionally show an expected result.
> 
> I don't think so. I can see the same problem without `correctFormat`:
> 
> * 11.1 [jsfiddle.net/warpech/wa58goL2/1](https://jsfiddle.net/warpech/wa58goL2/1/)
> * 12.0 [jsfiddle.net/warpech/xjrf0y1p/1](https://jsfiddle.net/warpech/xjrf0y1p/1/)
> 
> I think the problem is that Pikaday transforms the input before closing the editor. On this line, Pikaday chooses `defaultDateFormat`:
> 
> https://github.com/handsontable/handsontable/blob/6a1706061163b8d1f7752e54ef6f10efbc764b8b/handsontable/src/editors/dateEditor/dateEditor.js#L242
> 
> Why is this a problem starting from 12.0, I don't know. Maybe something has changed in the editor closing code.

I'll check that.

**Comment by wszymanski** [comment]
It seems that there are two bugs:

1. The value inside the cell is changed right after pressing `ENTER` key (from `01/07/2022` to `07/01/2022`).

https://user-images.githubusercontent.com/141330/188672567-668b8c41-0b33-4b2d-83c7-37dccf31029d.mp4

3. The value shown in the widget is wrong (it shows 7th of January 2022 instead of 1st of July 2022).

<img width="265" alt="Screenshot 2022-09-06 at 17 14 01" src="https://user-images.githubusercontent.com/141330/188671961-1514f079-d8a8-4040-9c36-321150467afd.png">

The first problem has been fixed by commit: https://github.com/handsontable/handsontable/commit/824ab4e9167a817d35857701190d4f753a86bed0. The `window` object is capturing an event later that `document.documentElement`. That's why `Pikaday` has been catching the `change` event as first (before `keydown` event). This leads us to some problems.

The second problem is investigated. Pikaday uses `moment` internally and call `toDate` method which doesn't work as I assume (please see the below example).

```
const momentDate = moment('01/07/2022', 'MM/DD/YYYY');
momentDate.toDate(); // Fri Jan 07 2022 00:00:00 GMT+0100 (Central European Standard Time)
```

**Comment by wszymanski** [comment]
I've found extra problem related to flickering while entering data by copy+paste and `enter` key.

https://user-images.githubusercontent.com/141330/188913217-bd99620b-4f61-4b7d-a59e-835a0c70316e.mp4

I'll restore mechanism which has been working up to version `12.0.0` (an extra `event.stopPropagation` blocked event captured by `Pikaday`).

**Comment by budnix** [comment]
@wszymanski, I believe you closed the issue by accident 😄  The issue should be closed by the support team after the fix is released and confirmed that it works.

**Comment by wszymanski** [comment]
> @wszymanski, I believe you closed the issue by accident 😄 The issue should be closed by the support team after the fix is released and confirmed that it works.

Sure, thanks. It has been closed automatically, after merging the PR. It probably [happened after using the word "fixed"](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

**Comment by aninde** [comment]
This issue will be fixed by v12.1.3

**Comment by AMBudnik** [comment]
Issue closed as fixed in Handsontable v12.1.3
